compiler/rustc_hir_analysis/src/check/mod.rs

```diff
@@ -546,3 +554,82 @@ fn bad_non_zero_sized_fields<'tcx>(
 pub fn potentially_plural_count(count: usize, word: &str) -> String {
     format!("{} {}{}", count, word, pluralize!(count))
 }
+
+pub fn check_function_signature<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mut cause: ObligationCause<'tcx>,
+    fn_id: DefId,
+    expected_sig: ty::PolyFnSig<'tcx>,
+) {
+    let local_id = fn_id.as_local().unwrap_or(CRATE_DEF_ID);
+
+    let param_env = ty::ParamEnv::empty();
+
+    let infcx = &tcx.infer_ctxt().build();
+    let ocx = ObligationCtxt::new(infcx);
+
+    let unnormalized_actual_sig = infcx.instantiate_binder_with_fresh_vars(
+        cause.span,
+        infer::HigherRankedType,
+        tcx.fn_sig(fn_id).instantiate_identity(),
+    );
+
+    let norm_cause = ObligationCause::misc(cause.span, local_id);
+    let actual_sig = ocx.normalize(&norm_cause, param_env, unnormalized_actual_sig);
+
+    let expected_sig = tcx.liberate_late_bound_regions(fn_id, expected_sig);
+
+    match ocx.eq(&cause, param_env, expected_sig, actual_sig) {
```

Comment:
This is not how to compare two binders. This is not very much different than using `sub`, which I think is still wrong. You should probably wrap both binders in a `FnPtr` and compare them directly. The error reporting will kinda suck, but that can be dealt with differently.

Reply:
The output changed from 
```
   = note: expected signature `fn(&PanicInfo<'_>) -> _`
              found signature `fn(&'static PanicInfo<'_>) -> _`
```
to
```
   = note: expected fn pointer `for<'a, 'b> fn(&'a PanicInfo<'b>) -> _`
              found fn pointer `for<'a> fn(&'static PanicInfo<'a>) -> _`
```

But only for lifetime errors.

For non-lifetime errors, it stayed expected/found signature, I believe that thanks to passing the signatures in `infer::ValuePairs` here (I'm not sure that's entirely correct)
```rust
err_ctxt.note_type_err(
    &mut diag,
    &cause,
    None,
    Some(infer::ValuePairs::Sigs(ExpectedFound {
        expected: tcx.liberate_late_bound_regions(fn_id, expected_sig),
        found: tcx.liberate_late_bound_regions(fn_id, actual_sig),
    })),
    err,
    false,
    false,
);
```